File: app/src/main/java/org/dhis2/usescases/main/program/ProgramModelHolder.java
package org.dhis2.usescases.main.program;

import android.content.res.Resources;
import android.graphics.drawable.Drawable;

import org.dhis2.R;
import org.dhis2.databinding.ItemProgramModelBinding;
import org.dhis2.utils.ColorUtils;
import org.dhis2.utils.Period;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.recyclerview.widget.RecyclerView;
import timber.log.Timber;

/**
 * QUADRAM. Created by ppajuelo on 13/06/2018.
 */

public class ProgramModelHolder extends RecyclerView.ViewHolder {

    private final ItemProgramModelBinding binding;

    public ProgramModelHolder(ItemProgramModelBinding binding) {
        super(binding.getRoot());
        this.binding = binding;
    }

    public void bind(ProgramContract.Presenter presenter, ProgramViewModel programViewModel) {
        binding.setProgram(programViewModel);
        binding.setPresenter(presenter);

        int color = ColorUtils.getColorFrom(programViewModel.color(), ColorUtils.getPrimaryColor(binding.programImage.getContext(), ColorUtils.ColorType.PRIMARY));
        int icon;
        if (programViewModel.icon() != null) {
            Resources resources = binding.programImage.getResources();
            String iconName = programViewModel.icon().startsWith("ic_") ? programViewModel.icon() : "ic_" + programViewModel.icon();
            icon = resources.getIdentifier(iconName, "drawable", binding.programImage.getContext().getPackageName());
        } else {
            icon = R.drawable.ic_program_default;
        }
        Drawable iconImage = AppCompatResources.getDrawable(binding.programImage.getContext(), R.drawable.ic_program_default);
        try {
            iconImage = AppCompatResources.getDrawable(binding.programImage.getContext(), icon);
        } catch (Exception e) {
            Timber.log(1, e);
        }

        if (iconImage != null) {
            iconImage.mutate();
        }

        binding.programImage.setImageDrawable(iconImage);
        binding.programImage.setColorFilter(ColorUtils.getContrastColor(color));
        binding.programImage.setBackgroundColor(color);

    }
}

File: app/src/main/java/org/dhis2/usescases/main/program/ProgramModelHolder.kt
package org.dhis2.usescases.main.program

import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.composethemeadapter.MdcTheme
import org.dhis2.R
import org.dhis2.commons.resources.ColorUtils
import org.dhis2.commons.resources.ResourceManager
import org.dhis2.commons.ui.MetadataIcon
import org.dhis2.commons.ui.MetadataIconData
import org.dhis2.databinding.ItemProgramModelBinding

class ProgramModelHolder(private val binding: ItemProgramModelBinding) :
    RecyclerView.ViewHolder(binding.root) {

    init {
        binding.composeProgramImage.setViewCompositionStrategy(
            ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed
        )
    }

    fun bind(presenter: ProgramPresenter, programViewModel: ProgramViewModel) {
        binding.program = programViewModel
        binding.presenter = presenter

        val color = ColorUtils.getColorFrom(
            programViewModel.color(),
            ColorUtils.getPrimaryColor(
                itemView.context,
                ColorUtils.ColorType.PRIMARY_LIGHT
            )
        )

        val iconResource = ResourceManager(itemView.context).getObjectStyleDrawableResource(
            programViewModel.icon(),
            R.drawable.ic_default_outline
        )

        binding.composeProgramImage.setContent {
            MdcTheme {
                MetadataIcon(
                    MetadataIconData(
                        programColor = color,
                        iconResource = iconResource
                    )
                )
            }
        }

        itemView.setOnClickListener {
            presenter.onItemClick(programViewModel)
        }

        binding.root.alpha = if (programViewModel.translucent()) {
            0.5f
        } else {
            1.0f
        }
    }
}
